src/test/java/com/ea/restaurant/controller/UserControllerE2ETest.java

```diff
@@ -0,0 +1,146 @@
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Objects;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.boot.test.web.client.TestRestTemplate;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.jdbc.core.JdbcTemplate;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
+public class UserControllerE2ETest {
+
+  @Autowired private JdbcTemplate jdbcTemplate;
+
+  @Autowired private TestRestTemplate restTemplate;
+
+  private User user1;
+  private User user2;
+
+  @BeforeEach
+  void setUp() {
+    jdbcTemplate.execute("TRUNCATE TABLE users CASCADE");
+    jdbcTemplate.execute("ALTER SEQUENCE users_id_seq RESTART");
+
+    user1 = UserFixture.buildUser();
+    user2 = UserFixture.buildUser();
+  }
+
+  @Test
+  void WhenControllerCreate_shouldCreateUser() {
+
+    var userSavedExpected = UserFixture.buildUser(1L);
+
+    HttpEntity<User> request = new HttpEntity<>(this.user1);
+    ResponseEntity<User> createResponse =
+        this.restTemplate.postForEntity("/users", request, User.class);
+    Assertions.assertEquals(HttpStatus.CREATED, createResponse.getStatusCode());
+    Assertions.assertEquals(
+        userSavedExpected.getId(), Objects.requireNonNull(createResponse.getBody()).getId());
+  }
+
+  @Test
+  void WhenControllerGetById_ShouldReturnExpectedUser() {
+
+    HttpEntity<User> request = new HttpEntity<>(this.user1);
+    ResponseEntity<User> createResponse =
+        this.restTemplate.postForEntity("/users", request, User.class);
+    ResponseEntity<User> getByIdResponse =
+        restTemplate.getForEntity(
+            "/users/{id}",
+            User.class,
+            Map.of("id", Objects.requireNonNull(createResponse.getBody()).getId()));
+    Assertions.assertEquals(HttpStatus.OK, getByIdResponse.getStatusCode());
+    Assertions.assertEquals(
+        Objects.requireNonNull(createResponse.getBody()).getId(),
+        Objects.requireNonNull(getByIdResponse.getBody()).getId());
+  }
+
+  @Test
+  void WhenControllerGetAll_ShouldReturnAllUsers() {
+    HttpEntity<User> request1 = new HttpEntity<>(this.user1);
+    HttpEntity<User> request2 = new HttpEntity<>(this.user2);
+    ResponseEntity<User> createResponse1 =
+        this.restTemplate.postForEntity("/users", request1, User.class);
+    ResponseEntity<User> createResponse2 =
+        this.restTemplate.postForEntity("/users", request2, User.class);
+    var usersSaved = Arrays.asList(createResponse1.getBody(), createResponse2.getBody());
+    ResponseEntity<User[]> getAllResponse = restTemplate.getForEntity("/users", User[].class);
```

Comment:
For all Objects of objects use `ParameterizedTypeReference`, to avoid User[].class and take advantage of java collections

Reply:
Done.